pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+      instances of user-defined classes also freeze their class
+      and its attributes).
+    * The ``__bases__`` chain in classes (freezing a class freezes its
+      base classes).
+
+5. Attempting to mutate a frozen object raises an exception (``MutationError``).
+
+
+Default (Im)Mutabiliy 
+---------------------
+
+- TODO: describe what classes create immutable objects
+
+- Strings, numbers, and tuples of immutable objects create immutable
+  objects by default
+
+
+Consequences of Deep Freezing
+=============================
+
+* Class Freezing: Freezing an instance of a user-defined class
+  will also freeze its class, potentially affecting all instances
+  of that class.
+* Metaclass Freezing: Since class objects have metaclasses,
+  freezing a class may propagate upwards through the metaclass
+  hierarchy.
+* Global State Impact: Freezing an object that references global
+  state (e.g., ``sys.modules``, built-in types) could inadvertently
+  freeze critical parts of the interpreter.
+
+As the above list shows, a side-effect of freezing an object is
+that its type becomes frozen too. Consider the following program,
+which is not legal in this PEP because it modifies the type of an
+immutable object:
+
+.. code-block:: c
+
+	class Counter:
+	  def __init__(self, initial_value):
+	    self.value = initial_value
+	  def inc(self):
+	    self.value += 1
+	  def dec(self):
+	    self.value -= 1
+	  def get(self):
+	    return self.value
+
+	c = Counter(0)
+	c.get() # returns 0 
+	freeze(c) # (1)
+	... 
+	Counter.get = lambda self: 42 # OK in CPython 3.12, throws exception with this PEP
+	c.get() # returns 42, even though c was frozen when the value was 0!
+
+With this PEP, the code above throws an exception on
+Line (1) because the type object for the ``Counter`` type
+is immutable. Our freeze algorithm takes care of this as
+it follows the class reference from ``c``. If we did not
+freeze the ``Counter`` type object, the above code would
+work and the counter will *appear* to be mutable because
+of the change to its class. 
+
+The dangers of not freezing the type is apparent when considering
+avoiding data races in a concurrent program. If a frozen counter
+is shared between two threads, the threads are still able to
+race on the ``Counter`` class type object.
+
+As types are frozen, this problem is avoided. Note that
+freezing a class needs to freeze its superclasses as well.
+
+
+Subclassing Immutable Classes
+-----------------------------
+
+CPython classes hold references to their subclasses.
+If immutability it taken literally, it would not be
+permitted to create a subclass of an immutable type.
+Because this reference is "accidental" and does not
+get exposed to the programmer in any dangerous way,
+we permit frozen classes to be subclassed (by mutable
+classes).
+
+
+Implementation Details
+======================
+
+1. Introduce an ``is_immutable`` flag in PyObject to indicate whether an
```

Comment:
Not a flag, an API call `_Py_ISIMMUTABLE` at the C API level and the `isimmutable` builtin